base/compiler/tfuncs.jl

```diff
@@ -724,20 +724,30 @@ function apply_type_tfunc(@nospecialize(headtypetype), @nospecialize args...)
     elseif isconstType(headtypetype)
         headtype = headtypetype.parameters[1]
     else
-        return Any
+        return Type
     end
     largs = length(args)
     if headtype === Union
         largs == 0 && return Const(Bottom)
-        largs == 1 && return args[1]
+        hasnonType = false
         for i = 1:largs
             ai = args[i]
-            if !isa(ai, Const) || !isa(ai.val, Type)
+            if isa(ai, Const)
+                if !isa(ai.val, Type)
+                    return Union{}
+                end
+            else
```

Comment:
Checking explicitly for if isa TypeVar might be necessary (And PartialTypeVar)

Reply:
yes. or `ai` could be a PartialTypeVar

Reply:
`ai` being a `PartialTypeVar` should be handled below. Do you mean `ai.val`?